src/Symfony/Component/FeatureToggle/FeatureCollection.php

```diff
@@ -12,43 +12,33 @@
 namespace Symfony\Component\FeatureToggle;
 
 use Psr\Container\ContainerInterface;
+use Symfony\Component\FeatureToggle\Provider\InMemoryProvider;
+use Symfony\Component\FeatureToggle\Provider\ProviderInterface;
 use function array_key_exists;
-use function array_shift;
-use function is_callable;
+use function array_merge;
 
-/** @implements \IteratorAggregate<int, Feature> */
-final class FeatureCollection implements ContainerInterface, \IteratorAggregate
+final class FeatureCollection implements ContainerInterface
 {
     /** @var array<string, Feature> */
     private array $features = [];
 
-    /** @var array<iterable<Feature>|(\Closure(): iterable<Feature>)> */
-    private array $featureProviders = [];
+    /** @var iterable<int, ProviderInterface> */
+    private iterable $providers;
 
     /**
-     * @param iterable<Feature> $features
+     * @param iterable<ProviderInterface> $providers
      */
-    public function __construct(iterable $features)
+    public function __construct(iterable $providers = [])
     {
-        $this->append($features);
+        $this->providers = $providers;
     }
 
     /**
-     * @param iterable<Feature>|(\Closure(): iterable<Feature>) $features
+     * @param list<Feature> $features
      */
-    private function append(iterable|\Closure $features): void
+    public static function withFeatures(array $features): self
```

Comment:
```suggestion
    public static function withFeatures(Feature ...$features): self
```

Reply:
This makes it much harder to add more parameters in the future in a backwards compatible way. Let's stick with array.

Reply:
it does not only makes it harder. It makes it impossible (unless we want to put them at the beginning of the signature, with a complex BC layer, but if would be weird if `$features` was not the first argument).

Reply:
It does not make much sense to add other parameters to a function named `withFeatures`, for BC then it would be better to add another method and deprecate this one instead.

Reply:
I don't see real benefits here. It makes the call to that method less intuitive, for something static analysis can do. Even if people don't use it, it will break somewhere down the road. No need to enforce anything here just for the sake of it IMO